File: src/exploration.c
/**
 * Evocore Exploration Control Implementation
 *
 * Implements dynamic exploration vs exploitation strategies.
 */

#define _GNU_SOURCE
#include "evocore/exploration.h"
#include "evocore/log.h"
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <math.h>

/*========================================================================
 * Constants
 *========================================================================*/

#define DEFAULT_MIN_RATE 0.01
#define DEFAULT_MAX_RATE 1.0
#define DEFAULT_DECAY_RATE 0.98
#define DEFAULT_TEMPERATURE 100.0
#define DEFAULT_COOLING_RATE 0.95
#define DEFAULT_UCB_C 1.4142  /* sqrt(2) */

#define STAGNATION_THRESHOLD 50
#define BOOST_FACTOR 2.0
#define BOOST_DURATION 10

/*========================================================================
 * Exploration Management
 *========================================================================*/

evocore_exploration_t* evocore_exploration_create(
    evocore_explore_strategy_t strategy,
    double base_rate
) {
    if (base_rate < 0.0 || base_rate > 1.0) return NULL;

    evocore_exploration_t *exp = calloc(1, sizeof(evocore_exploration_t));
    if (!exp) return NULL;

    exp->strategy = strategy;
    exp->base_rate = base_rate;
    exp->current_rate = base_rate;
    exp->min_rate = DEFAULT_MIN_RATE;
    exp->max_rate = DEFAULT_MAX_RATE;
    exp->decay_rate = DEFAULT_DECAY_RATE;
    exp->temperature = DEFAULT_TEMPERATURE;
    exp->cooling_rate = DEFAULT_COOLING_RATE;
    exp->ucb_c = DEFAULT_UCB_C;

    exp->best_fitness = -INFINITY;
    exp->recent_best = -INFINITY;
    exp->stagnation_count = 0;
    exp->total_evaluations = 0;

    exp->start_time = time(NULL);

    return exp;
}

void evocore_exploration_free(evocore_exploration_t *exp) {
    free(exp);
}

void evocore_exploration_reset(evocore_exploration_t *exp) {
    if (!exp) return;

    exp->current_rate = exp->base_rate;
    exp->best_fitness = -INFINITY;
    exp->recent_best = -INFINITY;
    exp->stagnation_count = 0;
    exp->total_evaluations = 0;
    exp->temperature = DEFAULT_TEMPERATURE;
    exp->start_time = time(NULL);
}

void evocore_exploration_set_bounds(
    evocore_exploration_t *exp,
    double min_rate,
    double max_rate
) {
    if (!exp) return;
    exp->min_rate = fmin(1.0, fmax(0.0, min_rate));
    exp->max_rate = fmin(1.0, fmax(0.0, max_rate));
}

void evocore_exploration_set_decay_rate(
    evocore_exploration_t *exp,
    double decay_rate
) {
    if (!exp) return;
    exp->decay_rate = fmin(1.0, fmax(0.0, decay_rate));
}

void evocore_exploration_set_temperature(
    evocore_exploration_t *exp,
    double temperature,
    double cooling_rate
) {
    if (!exp) return;
    exp->temperature = fmax(0.001, temperature);
    exp->cooling_rate = fmin(1.0, fmax(0.0, cooling_rate));
}

void evocore_exploration_set_ucb_c(
    evocore_exploration_t *exp,
    double ucb_c
) {
    if (!exp) return;
    exp->ucb_c = fmax(0.0, ucb_c);
}

/*========================================================================
 * Rate Updates
 *========================================================================*/

double evocore_exploration_update(
    evocore_exploration_t *exp,
    size_t generation,
    double best_fitness
) {
    if (!exp) return 0.0;

    exp->total_evaluations++;

    /* Update best fitness tracking */
    bool improved = false;
    if (best_fitness > exp->best_fitness) {
        exp->best_fitness = best_fitness;
        improved = true;
    }

    /* Update strategy-specific rate */
    switch (exp->strategy) {
        case EVOCORE_EXPLORE_FIXED:
            /* Constant rate */
            exp->current_rate = exp->base_rate;
            break;

        case EVOCORE_EXPLORE_DECAY: {
            /* Exponential decay: rate = base * decay^gen */
            double decayed = exp->base_rate * pow(exp->decay_rate, (double)generation);
            exp->current_rate = fmax(exp->min_rate, decayed);
            break;
        }

        case EVOCORE_EXPLORE_ADAPTIVE: {
            /* Increase on stagnation, decrease on improvement */
            if (improved) {
                exp->stagnation_count = 0;
                /* Decrease exploration when improving */
                exp->current_rate *= 0.9;
            } else {
                exp->stagnation_count++;
                /* Increase exploration when stagnant */
                if (exp->stagnation_count > STAGNATION_THRESHOLD / 2) {
                    exp->current_rate *= 1.1;
                }
            }
            break;
        }

        case EVOCORE_EXPLORE_UCB1:
            /* Rate decreases as we gather more data */
            if (exp->total_evaluations > 0) {
                exp->current_rate = exp->ucb_c / sqrt((double)exp->total_evaluations);
            }
            break;

        case EVOCORE_EXPLORE_BOLTZMANN: {
            /* Based on temperature */
            exp->current_rate = exp->temperature / DEFAULT_TEMPERATURE;
            break;
        }
    }

    /* Clamp to bounds */
    exp->current_rate = fmax(exp->min_rate,
                            fmin(exp->max_rate, exp->current_rate));

    /* Cool temperature for Boltzmann */
    if (exp->strategy == EVOCORE_EXPLORE_BOLTZMANN) {
        exp->temperature = evocore_cool_temperature(exp->temperature, exp->cooling_rate);
    }

    return exp->current_rate;
}

double evocore_exploration_get_rate(const evocore_exploration_t *exp) {
    return exp ? exp->current_rate : 0.0;
}

bool evocore_exploration_should_explore(
    const evocore_exploration_t *exp,
    unsigned int *seed
) {
    if (!exp) return false;
    if (exp->current_rate <= 0.0) return false;
    if (exp->current_rate >= 1.0) return true;

    double r = (double)rand_r(seed) / (double)RAND_MAX;
    return r < exp->current_rate;
}

/*========================================================================
 * Bandit Selection (UCB1)
 *========================================================================*/

evocore_bandit_t* evocore_bandit_create(size_t arm_count, double ucb_c) {
    if (arm_count == 0) return NULL;

    evocore_bandit_t *bandit = calloc(1, sizeof(evocore_bandit_t));
    if (!bandit) return NULL;

    bandit->arms = calloc(arm_count, sizeof(evocore_bandit_arm_t));
    if (!bandit->arms) {
        free(bandit);
        return NULL;
    }

    bandit->count = arm_count;
    bandit->ucb_c = ucb_c;
    bandit->total_pulls = 0;

    return bandit;
}

void evocore_bandit_free(evocore_bandit_t *bandit) {
    if (!bandit) return;
    free(bandit->arms);
    free(bandit);
}

size_t evocore_bandit_select_ucb(const evocore_bandit_t *bandit) {
    if (!bandit || bandit->count == 0) return 0;

    size_t best_arm = 0;
    double best_ucb = -INFINITY;

    for (size_t i = 0; i < bandit->count; i++) {
        const evocore_bandit_arm_t *arm = &bandit->arms[i];

        double ucb;
        if (arm->count == 0) {
            /* Never pulled, select it */
            ucb = INFINITY;
        } else {
            /* UCB1 formula: mean + c * sqrt(ln(n) / n_i) */
            double exploration = bandit->ucb_c *
                               sqrt(log((double)bandit->total_pulls) / (double)arm->count);
            ucb = arm->mean_reward + exploration;
        }

        if (ucb > best_ucb) {
            best_ucb = ucb;
            best_arm = i;
        }
    }

    return best_arm;
}

void evocore_bandit_update(evocore_bandit_t *bandit, size_t arm_idx, double reward) {
    if (!bandit || arm_idx >= bandit->count) return;

    evocore_bandit_arm_t *arm = &bandit->arms[arm_idx];

    arm->count++;
    arm->total_reward += reward;
    arm->mean_reward = arm->total_reward / (double)arm->count;

    bandit->total_pulls++;
}

size_t evocore_bandit_arm_count(const evocore_bandit_t *bandit) {
    return bandit ? bandit->count : 0;
}

bool evocore_bandit_get_stats(
    const evocore_bandit_t *bandit,
    size_t arm_idx,
    size_t *out_count,
    double *out_mean
) {
    if (!bandit || arm_idx >= bandit->count) return false;

    const evocore_bandit_arm_t *arm = &bandit->arms[arm_idx];

    if (out_count) *out_count = arm->count;
    if (out_mean) *out_mean = arm->mean_reward;

    return true;
}

void evocore_bandit_reset(evocore_bandit_t *bandit) {
    if (!bandit) return;

    memset(bandit->arms, 0, bandit->count * sizeof(evocore_bandit_arm_t));
    bandit->total_pulls = 0;
}

/*========================================================================
 * Boltzmann Selection
 *========================================================================*/

size_t evocore_boltzmann_select(
    const double *values,
    size_t count,
    double temperature,
    unsigned int *seed
) {
    if (!values || count == 0) return 0;
    if (temperature < 0.001) {
        /* Very low temperature: select max */
        size_t best = 0;
        double best_val = values[0];
        for (size_t i = 1; i < count; i++) {
            if (values[i] > best_val) {
                best_val = values[i];
                best = i;
            }
        }
        return best;
    }

    /* Calculate Boltzmann probabilities: p_i = exp(v_i / T) / sum(exp(v_j / T)) */
    double *probs = calloc(count, sizeof(double));
    if (!probs) return 0;

    double sum = 0.0;
    double max_val = values[0];
    for (size_t i = 1; i < count; i++) {
        if (values[i] > max_val) max_val = values[i];
    }

    /* Subtract max for numerical stability */
    for (size_t i = 0; i < count; i++) {
        probs[i] = exp((values[i] - max_val) / temperature);
        sum += probs[i];
    }

    if (sum < 0.0001) {
        /* All values are the same or very low temperature */
        free(probs);
        return (size_t)((double)rand_r(seed) / (double)RAND_MAX * count);
    }

    /* Normalize and select */
    double r = (double)rand_r(seed) / (double)RAND_MAX * sum;
    double cumulative = 0.0;
    size_t selected = 0;

    for (size_t i = 0; i < count; i++) {
        cumulative += probs[i];
        if (r <= cumulative) {
            selected = i;
            break;
        }
    }

    free(probs);
    return selected;
}

double evocore_cool_temperature(double temperature, double cooling_rate) {
    return fmax(0.001, temperature * cooling_rate);
}

/*========================================================================
 * Adaptive Exploration
 *========================================================================*/

bool evocore_exploration_is_stagnant(
    const evocore_exploration_t *exp,
    size_t threshold
) {
    if (!exp) return false;
    return exp->stagnation_count >= threshold;
}

void evocore_exploration_boost(
    evocore_exploration_t *exp,
    double factor
) {
    if (!exp) return;

    exp->current_rate *= factor;
    exp->current_rate = fmin(exp->max_rate, exp->current_rate);
}

double evocore_exploration_improvement_rate(const evocore_exploration_t *exp) {
    if (!exp || exp->total_evaluations == 0) return 0.0;

    /* Improvement per evaluation (simplified) */
    if (isinf(exp->best_fitness) || exp->best_fitness == -INFINITY) {
        return 0.0;
    }

    return exp->best_fitness / (double)exp->total_evaluations;
}

File: python/evocore/strategy/exploration.py
"""
Exploration control for evocore Python bindings.

Provides exploration rate scheduling and multi-armed bandit algorithms.
"""

from typing import Optional, List, Tuple
from enum import IntEnum
import numpy as np
from ..utils.error import EvocoreError


class ExploreStrategy(IntEnum):
    """Exploration strategy types."""
    FIXED = 0
    DECAY = 1
    ADAPTIVE = 2
    UCB1 = 3
    BOLTZMANN = 4


class Exploration:
    """
    Exploration rate controller.

    Manages exploration/exploitation balance with various strategies.

    Example:
        >>> exp = Exploration(ExploreStrategy.ADAPTIVE, base_rate=0.5)
        >>> exp.set_bounds(0.1, 0.9)
        >>> rate = exp.update(generation=10, best_fitness=0.8)
    """

    __slots__ = ('_exp', '_ffi', '_lib')

    def __init__(self, strategy: ExploreStrategy, base_rate: float = 0.5,
                 *, _raw: bool = False):
        """
        Initialize exploration controller.

        Args:
            strategy: Exploration strategy to use
            base_rate: Initial exploration rate
            _raw: Internal flag for alternative construction
        """
        if _raw:
            self._exp = None
            self._ffi = None
            self._lib = None
            return

        from .._evocore import ffi, lib
        self._ffi = ffi
        self._lib = lib

        self._exp = lib.evocore_exploration_create(int(strategy), base_rate)
        if self._exp == ffi.NULL:
            raise EvocoreError("Failed to create exploration controller")

    def __del__(self):
        """Clean up exploration controller."""
        if hasattr(self, '_exp') and self._exp is not None:
            self._lib.evocore_exploration_free(self._exp)

    @property
    def strategy(self) -> ExploreStrategy:
        """Current exploration strategy."""
        return ExploreStrategy(self._exp.strategy)

    @property
    def current_rate(self) -> float:
        """Current exploration rate."""
        return self._lib.evocore_exploration_get_rate(self._exp)

    @property
    def base_rate(self) -> float:
        """Base exploration rate."""
        return self._exp.base_rate

    def set_bounds(self, min_rate: float, max_rate: float) -> None:
        """
        Set rate bounds.

        Args:
            min_rate: Minimum exploration rate
            max_rate: Maximum exploration rate
        """
        self._lib.evocore_exploration_set_bounds(self._exp, min_rate, max_rate)

    def set_decay_rate(self, decay_rate: float) -> None:
        """
        Set decay rate for DECAY strategy.

        Args:
            decay_rate: Decay factor per generation
        """
        self._lib.evocore_exploration_set_decay_rate(self._exp, decay_rate)

    def set_temperature(self, temperature: float, cooling_rate: float) -> None:
        """
        Set temperature parameters for BOLTZMANN strategy.

        Args:
            temperature: Initial temperature
            cooling_rate: Cooling rate per step
        """
        self._lib.evocore_exploration_set_temperature(self._exp, temperature, cooling_rate)

    def set_ucb_c(self, ucb_c: float) -> None:
        """
        Set UCB constant for UCB1 strategy.

        Args:
            ucb_c: UCB exploration constant
        """
        self._lib.evocore_exploration_set_ucb_c(self._exp, ucb_c)

    def update(self, generation: int, best_fitness: float) -> float:
        """
        Update exploration rate based on progress.

        Args:
            generation: Current generation
            best_fitness: Best fitness achieved

        Returns:
            New exploration rate
        """
        return self._lib.evocore_exploration_update(self._exp, generation, best_fitness)

    def should_explore(self, seed: Optional[int] = None) -> bool:
        """
        Randomly decide whether to explore.

        Args:
            seed: Optional random seed

        Returns:
            True if should explore
        """
        seed_ptr = self._ffi.new("unsigned int *", seed if seed is not None else 0)
        return self._lib.evocore_exploration_should_explore(self._exp, seed_ptr)

    def is_stagnant(self, threshold: int = 10) -> bool:
        """
        Check if evolution is stagnant.

        Args:
            threshold: Generations without improvement

        Returns:
            True if stagnant
        """
        return self._lib.evocore_exploration_is_stagnant(self._exp, threshold)

    def boost(self, factor: float = 1.5) -> None:
        """
        Boost exploration rate temporarily.

        Args:
            factor: Multiplier for exploration rate
        """
        self._lib.evocore_exploration_boost(self._exp, factor)

    def improvement_rate(self) -> float:
        """
        Get improvement rate.

        Returns:
            Rate of improvement
        """
        return self._lib.evocore_exploration_improvement_rate(self._exp)

    def reset(self) -> None:
        """Reset to initial state."""
        self._lib.evocore_exploration_reset(self._exp)

    def __repr__(self) -> str:
        return (f"Exploration(strategy={self.strategy.name}, "
                f"rate={self.current_rate:.3f})")


class Bandit:
    """
    Multi-armed bandit for action selection.

    Uses UCB1 algorithm to balance exploration and exploitation.

    Example:
        >>> bandit = Bandit(arm_count=4, ucb_c=2.0)
        >>> arm = bandit.select()
        >>> bandit.update(arm, reward=0.8)
    """

    __slots__ = ('_bandit', '_ffi', '_lib')

    def __init__(self, arm_count: int, ucb_c: float = 2.0, *, _raw: bool = False):
        """
        Initialize multi-armed bandit.

        Args:
            arm_count: Number of arms
            ucb_c: UCB exploration constant
            _raw: Internal flag for alternative construction
        """
        if _raw:
            self._bandit = None
            self._ffi = None
            self._lib = None
            return

        from .._evocore import ffi, lib
        self._ffi = ffi
        self._lib = lib

        self._bandit = lib.evocore_bandit_create(arm_count, ucb_c)
        if self._bandit == ffi.NULL:
            raise EvocoreError("Failed to create bandit")

    def __del__(self):
        """Clean up bandit."""
        if hasattr(self, '_bandit') and self._bandit is not None:
            self._lib.evocore_bandit_free(self._bandit)

    @property
    def arm_count(self) -> int:
        """Number of arms."""
        return self._lib.evocore_bandit_arm_count(self._bandit)

    @property
    def total_pulls(self) -> int:
        """Total number of arm pulls."""
        return self._bandit.total_pulls

    def select(self) -> int:
        """
        Select an arm using UCB1.

        Returns:
            Selected arm index
        """
        return self._lib.evocore_bandit_select_ucb(self._bandit)

    def update(self, arm: int, reward: float) -> None:
        """
        Update arm statistics after observing reward.

        Args:
            arm: Arm index
            reward: Observed reward
        """
        self._lib.evocore_bandit_update(self._bandit, arm, reward)

    def get_stats(self, arm: int) -> Tuple[int, float]:
        """
        Get statistics for an arm.

        Args:
            arm: Arm index

        Returns:
            Tuple of (pull_count, mean_reward)
        """
        count = self._ffi.new("size_t *")
        mean = self._ffi.new("double *")

        success = self._lib.evocore_bandit_get_stats(self._bandit, arm, count, mean)

        if not success:
            return (0, 0.0)

        return (count[0], mean[0])

    def reset(self) -> None:
        """Reset all arm statistics."""
        self._lib.evocore_bandit_reset(self._bandit)

    def __repr__(self) -> str:
        return f"Bandit(arms={self.arm_count}, pulls={self.total_pulls})"


def boltzmann_select(values: np.ndarray, temperature: float,
                     seed: Optional[int] = None) -> int:
    """
    Select index using Boltzmann distribution.

    Args:
        values: Array of values to select from
        temperature: Temperature parameter
        seed: Optional random seed

    Returns:
        Selected index
    """
    from .._evocore import ffi, lib

    values = np.asarray(values, dtype=np.float64)
    values_arr = ffi.new("double[]", values.tolist())
    seed_ptr = ffi.new("unsigned int *", seed if seed is not None else 0)

    return lib.evocore_boltzmann_select(values_arr, len(values), temperature, seed_ptr)


def cool_temperature(temperature: float, cooling_rate: float) -> float:
    """
    Apply temperature cooling.

    Args:
        temperature: Current temperature
        cooling_rate: Cooling rate

    Returns:
        New temperature
    """
    from .._evocore import lib
    return lib.evocore_cool_temperature(temperature, cooling_rate)


__all__ = [
    'ExploreStrategy',
    'Exploration',
    'Bandit',
    'boltzmann_select',
    'cool_temperature',
]
